spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -150,8 +167,54 @@ else if (reply instanceof AbstractIntegrationMessageBuilder) {
 			}
 		}
 
-		Message<?> replyMessage = createOutputMessage(reply, requestHeaders);
-		sendOutput(replyMessage, replyChannel);
+		if (this.asyncReplySupported && reply instanceof ListenableFuture<?>) {
+			ListenableFuture<?> future = (ListenableFuture<?>) reply;
+			final Object theReplyChannel = replyChannel;
+			future.addCallback(new ListenableFutureCallback<Object>() {
+
+				@Override
+				public void onSuccess(Object result) {
+					try {
+						sendOutput(createOutputMessage(result, requestHeaders), theReplyChannel, false);
+					}
+					catch (Exception e) {
+						Exception exceptionToLog = e;
+						if (!(e instanceof MessagingException)) {
+							exceptionToLog = new MessageHandlingException(requestMessage, e);
+						}
+						logger.error("Failed to send async reply: " + result.toString(), exceptionToLog);
```

Comment:
Can you explain why don't you use `onFailure()` here?
I understand the `future.addCallback()` premise like a just notification about future changes.
But we should be consistent with existing SI flow behavior algorithm and really try to "handle async error" .
Just log the problem from the donwstream isn't good at all.

I argue that we have to send `ErrorMessage` to the `errorChannel` from headers for this `sendOutput()` problem.
Just because there is no any other hooks to react to the `ListenableFuture` value asynchronously.